File: langs/dotnet/src/Dotlyte/SchemaViolation.cs
namespace Dotlyte;

/// <summary>
/// Represents a single schema validation violation.
/// </summary>
/// <param name="Key">The config key that violated the rule.</param>
/// <param name="Message">Human-readable description of the violation.</param>
/// <param name="Rule">The schema rule that was violated.</param>
public sealed record SchemaViolation(string Key, string Message, SchemaRule Rule)
{
    /// <summary>String representation.</summary>
    public override string ToString() => $"{Key}: {Message}";
}

File: langs/php/src/SchemaViolation.php
<?php

declare(strict_types=1);

namespace Dotlyte;

/**
 * Represents a single schema violation.
 */
final class SchemaViolation
{
    public function __construct(
        public readonly string $key,
        public readonly string $message,
        public readonly string $rule,
    ) {}

    public function __toString(): string
    {
        return "[{$this->rule}] {$this->key}: {$this->message}";
    }
}
